lib/puma/client.rb

```diff
@@ -500,29 +500,29 @@ def read_body
       # after this
       remain = @body_remain
 
-      if remain > CHUNK_SIZE
-        want = CHUNK_SIZE
-      else
-        want = remain
-      end
+      # don't bother with reading zero bytes
+      unless remain.zero?
 
-      begin
-        chunk = @io.read_nonblock(want, @read_buffer)
-      rescue IO::WaitReadable
-        return false
-      rescue SystemCallError, IOError
-        raise ConnectionError, "Connection error detected during read"
-      end
+        want = remain > CHUNK_SIZE ? CHUNK_SIZE : remain
```

Comment:
Optional Suggestion: We can also remove this variable assignment and move this logic inline later:

```ruby
        begin
          chunk = @io.read_nonblock(remain.clamp(0, CHUNK_SIZE), @read_buffer)
```

I feel like that's clearer and cleaner. I'm fine with the code as is, but would prefer `clamp` if you are willing to make that change.

Reply:
I can do that.